common/src/warp_runner/ui_adapter/mod.rs

```diff
@@ -211,6 +209,12 @@ pub async fn conversation_to_chat(
     .collect()
     .await;
 
+    let pinned: VecDeque<_> = messages
+        .iter()
+        .cloned()
```

Comment:
you probably want to call `.cloned()` after `.filter()` so that you don't clone every message in the conversation just to throw it away.